### PR TITLE
[TECHNICAL SUPPORT] LPS-116472 Implement upgrade process

### PR DESCRIPTION
**Pull request by rodrigopaulino**
Relevant ticket:
https://issues.liferay.com/browse/LPS-116472

I will run `ci:test:sf`, but it will fail due to https://issues.liferay.com/browse/LPS-116530

### CODE REVIEW

**Comment by rodrigopaulino** [comment]
ci:test:sf

**Comment by rodrigopaulino** [comment]
ci:test:relevant

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/resources/META-INF/portlet-model-hints.xml [hunk 1]
```diff
@@ -72,7 +72,7 @@
 			<validator name="required" />
 		</field>
 		<field localized="true" name="description" type="String">
-			<hint-collection name="TEXTAREA" />
+			<hint-collection name="CLOB" />
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/resources/META-INF/portlet-model-hints.xml:75]: Hi @rodrigopaulino, why did you decide to change the column's data type to CLOB? According to our discussion on [PTR-1777](https://issues.liferay.com/browse/PTR-1777), I remember that you decided to not do this. Can you please let me know why did you change your mind?

Thanks.

**Comment by natocesarrego** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/DDMServiceUpgrade.java [hunk 2]
```diff
@@ -230,8 +230,10 @@ public void register(Registry registry) {
 
 		registry.register("2.0.7", "2.0.8", new DummyUpgradeStep());
 
+		registry.register("2.0.8", "2.0.9", new DummyUpgradeStep());
+
 		registry.register(
-			"2.0.8", "3.0.0",
+			"2.0.9", "3.0.0",
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/DDMServiceUpgrade.java:236]: @rodrigopaulino, why was it necessary?

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/DDMServiceUpgrade.java [hunk 3]
```diff
@@ -287,8 +289,10 @@ public void register(Registry registry) {
 		registry.register(
 			"3.2.3", "3.2.4", new UpgradeDDMContent(_jsonFactory));
 
+		registry.register("3.2.4", "3.2.5", new DummyUpgradeStep());
+
 		registry.register(
-			"3.2.4", "3.3.0",
+			"3.2.5", "3.3.0",
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/DDMServiceUpgrade.java:295]: @rodrigopaulino, why was it necessary?

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v3_7_1/UpgradeSchema.java [hunk 4]
```diff
@@ -24,6 +26,23 @@
...
+			}
+
+			if (!hasColumnType(
+					DDMFormInstanceVersionTable.TABLE_NAME, "description",
+					"TEXT null")) {
+
+				alter(
+					DDMFormInstanceVersionTable.class,
+					new AlterColumnType("description", "TEXT null"));
+			}
```
Comment (natocesarrego) [inline modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v3_7_1/UpgradeSchema.java:45]: @rodrigopaulino, shouldn't this code change the column type to `CLOB`? I believe I didn't get. 😞

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/resources/META-INF/portlet-model-hints.xml (hunk 1 above)
Comment (rodrigopaulino) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/resources/META-INF/portlet-model-hints.xml:75]: Sure thing, @natocesarrego. Including a validation to limit the description field length is much more difficult than expected. Since we can't predict how much characters a translation is going to take, how many languages the user is going to use and if they plan to add more to the set of available languages in the future, there's no way to determine a significant size limit that can cover all cases. Also, changing the column to CLOB is, in fact, a micro change. You can check the full discussion about this decision with the Upgrade SME, Alberto Chaparro, in the [PTR-1804](https://issues.liferay.com/browse/PTR-1804).

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/DDMServiceUpgrade.java (hunk 2 above)
Comment (rodrigopaulino) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/DDMServiceUpgrade.java:236]: This is necessary every time an Upgrade needs to be backported. This reserves the version number that's going to be used on 7.1.x.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/DDMServiceUpgrade.java (hunk 3 above)
Comment (rodrigopaulino) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/DDMServiceUpgrade.java:295]: This is necessary every time an Upgrade needs to be backported. This reserves the version number that's going to be used on 7.2.x.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v3_7_1/UpgradeSchema.java (hunk 4 above)
Comment (rodrigopaulino) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/v3_7_1/UpgradeSchema.java:45]: Basically, Liferay translates "TEXT" to large character objects and it depends on the database you're using. When using, MySQL for example, it will be `LONGTEXT`. Using Oracle Database, it will be `CLOB`.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/DDMServiceUpgrade.java (hunk 3 above)
Comment (rodrigopaulino) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/DDMServiceUpgrade.java:295]: Practically, if someone is trying to upgrade from 7.2 to the latest, Liferay will be able to pick it up from their schema version onwards.

File: modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/DDMServiceUpgrade.java (hunk 2 above)
Comment (rodrigopaulino) [reply modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/upgrade/DDMServiceUpgrade.java:236]: Practically, if someone is trying to upgrade from 7.1 to the latest, Liferay will be able to pick it up from their schema version onwards.

**Comment by natocesarrego** [comment]
Thanks for your answers, @rodrigopaulino. 🙂 I've no more comments regarding this fix. Green light for you. 😄
